feat: Add workspace application support

kylecarbs opened the pull request:
This adds applications as a property to a workspace agent.

The resource is added to the Terraform provider here:
https://github.com/coder/terraform-provider-coder/pull/17

Apps will be opened in the dashboard or via the CLI
with `coder open <name>`. If `command` is specified, a
terminal will appear locally and on the web. If `target`
is specified, the browser will open to an exposed instance
of that target.

- [x] Fix odd context cancel bug on HTTP transport.

kylecarbs committed:
feat: Add app support

This adds apps as a property to a workspace agent.

The resource is added to the Terraform provider here:
https://github.com/coder/terraform-provider-coder/pull/17

Apps will be opened in the dashboard or via the CLI
with `coder open <name>`. If `command` is specified, a
terminal will appear locally and in the web. If `target`
is specified, the browser will open to an exposed instance
of that target.

kylecarbs committed:
Merge branch 'main' into devurls

kylecarbs committed:
Merge branch 'main' into devurls

kylecarbs committed:
Compare fields in apps test

kylecarbs committed:
Update Terraform provider to use relative path

kylecarbs committed:
Add some basic structure for routing

kylecarbs committed:
chore: Remove interface from coderd and lift API surface

Abstracting coderd into an interface added misdirection because
the interface was never intended to be fulfilled outside of a single
implementation.

This lifts the abstraction, and attaches all handlers to a root struct
named `*coderd.API`.

kylecarbs committed:
Merge branch 'routeclean' into devurls

kylecarbs committed:
Merge branch 'main' into devurls

kylecarbs committed:
Add basic proxy logic

kylecarbs committed:
Add proxying based on path

kylecarbs committed:
Merge branch 'main' into apps

kylecarbs committed:
Add app proxying for wildcards

kylecarbs committed:
Add wsconncache

kylecarbs committed:
fix: Race when writing to a closed pipe

This is such an intermittent race it's difficult to track,
but regardless this is an improvement to the code.

kylecarbs committed:
fix: Race when writing to a closed pipe

This is such an intermittent race it's difficult to track,
but regardless this is an improvement to the code.

kylecarbs committed:
fix: Race when writing to a closed pipe

This is such an intermittent race it's difficult to track,
but regardless this is an improvement to the code.

kylecarbs committed:
fix: Race when writing to a closed pipe

This is such an intermittent race it's difficult to track,
but regardless this is an improvement to the code.

kylecarbs committed:
Merge branch 'readclose' into apps

kylecarbs committed:
Merge branch 'readclose' into apps

kylecarbs committed:
Merge branch 'readclose' into apps

kylecarbs committed:
Merge branch 'readclose' into apps

spikecurtis commented on coderd/wsconncache/wsconncache.go:
```diff
@@ -0,0 +1,130 @@
+type Conn struct {
+	*agent.Conn
+
+	locks         atomic.Uint64
```
you increment and decrement this number, but never read it or use it for any conditional logic.

kylecarbs replied:
Yup, this is all WIP.

spikecurtis commented:
At a high level, I also find it a bit strange that we're having coderd negotiate a WebRTC connection to the agent, when the agent is _already_ connected to coderd via the channel that the ICE negotiation is happening over.

kylecarbs commented on coderd/wsconncache/wsconncache.go:
```diff
@@ -0,0 +1,130 @@
+			_ = conn.CloseWithError(xerrors.New("cache timeout"))
+		}()
+		c.connMutex.Lock()
+		c.conns[id] = conn
```
This is a good catch. Will certainly fix!

kylecarbs commented on coderd/coderd.go:
```diff
@@ -93,6 +97,15 @@ func New(options *Options) *API {
 		tracing.HTTPMW(api.TracerProvider, "coderd.http"),
 	)
 
+	r.Route("/{user}/{workspaceagent}/{application}", func(r chi.Router) {
```
Applications aren't accessed via the `/api/v2` route. They'll be served as if they're on the website, but really aren't.

eg. `/@kylecarbs/dev/apps/code-server`

spikecurtis replied:
I can see that.  My question is, why?  In particular, if they aren't on a versioned path, we may be painting our future selves into a corner

kylecarbs replied:
I don't think we should expose users to API routes. I can't think of any applications that redirect a client to an API URL to serve content.

This isn't really an API endpoint either, it's much more of an HTTP proxy.

kylecarbs commented on provisionersdk/proto/provisioner.proto:
```diff
@@ -80,12 +80,22 @@ message Agent {
+message App {
+    string name = 1;
+    string command = 2;
+    string url = 3;
```
Good catch.

spikecurtis commented:
> In a high availability deployment where multiple Coder replicas exist, there's no guarantee that the agent is connected to the dialed replica.

Ah, sure, but we have to establish a channel to exchange control messages.  So, before we can even start negotiating the WebRTC, we have a channel to the agent up.  Do we not want to use that channel in the hopes that WebRTC can find a path with fewer hops?

kylecarbs committed:
Add workspace route proxying endpoint

- Makes the workspace conn cache concurrency-safe
- Reduces unnecessary open checks in `peer.Channel`
- Fixes the use of a temporary context when dialing a workspace agent

kylecarbs committed:
Add embed errors

kylecarbs committed:
chore: Refactor site to improve testing

It was difficult to develop this package due to the
embed build tag being mandatory on the tests. The logic
to test doesn't require any embedded files.

kylecarbs committed:
Merge branch 'refactorsite' into apps

kylecarbs committed:
Add test for error handler

kylecarbs commented on provisionersdk/proto/provisioner.proto:
```diff
@@ -80,12 +80,22 @@ message Agent {
     oneof auth {
-        string token = 8;
-        string instance_id = 9;
+        string token = 9;
```
It's nicer structurally from my perspective.

kylecarbs commented:
We already do use that channel as the signal exchange for WebRTC. If we didn't, we could never establish a connection.

kylecarbs committed:
Remove unused access url

kylecarbs committed:
Add RBAC tests

kylecarbs committed:
Merge branch 'main' into apps

kylecarbs committed:
Merge branch 'main' into apps

vapurrmaid commented on coderd/coderd.go:
```diff
@@ -93,6 +96,20 @@ func New(options *Options) *API {
+	// %40 is the encoded character of the @ symbol. VS Code Web does
+	// not handle character encoding properly, so it's safe to assume
+	// other applications might not as well.
+	r.Route("/%40{user}/{workspacename}/apps/{workspaceapp}", apps)
```
Does this mean the FE should use `encodeURIComponent` ?

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/encodeURIComponent

kylecarbs replied:
Nah, we should always use `@` if possible. This is an edge-case for where certain client-side applications don't handle the `@` properly.

vapurrmaid commented on coderd/coderd.go:
```diff
@@ -93,6 +96,20 @@ func New(options *Options) *API {
+	// not handle character encoding properly, so it's safe to assume
+	// other applications might not as well.
+	r.Route("/%40{user}/{workspacename}/apps/{workspaceapp}", apps)
+	r.Route("/@{user}/{workspacename}/apps/{workspaceapp}", apps)
```
**Question**: Why do we have both then?

presleyp commented on site/site.go:
```diff
@@ -17,6 +19,15 @@ import (
+// metadata in error pages.
+func WithAPIResponse(ctx context.Context, apiResponse APIResponse) context.Context {
+	return context.WithValue(ctx, apiResponseContextKey{}, apiResponse)
+}
```
Is this dev url specific or is it related to the error message discussion from yesterday?

greyscaled reviewed (commented):
FE = ✔️

kylecarbs committed:
Fix dial agent syntax

kylecarbs committed:
Merge branch 'main' into apps

kylecarbs committed:
Fix linting errors

f0ssel commented on coderd/rbac/README.md:
```diff
@@ -51,7 +51,7 @@ This can be represented by the following truth table, where Y represents *positi
 
 ## Example Permissions
 
-- `+site.*.*.read`: allowed to perform the `read` action against all objects of type `devurl` in a given Coder deployment.
```
kill it with 🔥

f0ssel commented on site/src/api/typesGenerated.ts:
```diff
@@ -420,6 +421,15 @@ export interface WorkspaceAgentResourceMetadata {
+  readonly name: string
+  readonly command?: string
+  readonly access_url?: string
+  readonly icon: string
```
Is icon required? I think we should have some fallback on the frontend and make it optional in terraform if it isn't already.

kylecarbs replied:
True. I shall fix

f0ssel reviewed (approved):
Generally everything looks good to me. A little long so I didn't dive in super hard but the tests seem to cover enough bases to give me confidence.

kylecarbs committed:
Fix gen

kylecarbs committed:
Fix icon required

kylecarbs committed:
Merge branch 'main' into apps

kylecarbs committed:
Adjust migration number

kylecarbs committed:
Fix proxy error status code

kylecarbs committed:
Fix empty db lookup

kylecarbs committed:
Merge branch 'main' into apps

kylecarbs committed:
Merge branch 'main' into apps
